Feat: saved scripts (#14409)
fleetdm/fleet: docs/REST API/rest-api.md

#### Parameters

| Name            | Type    | In   | Description                                                                                    |
| ----            | ------- | ---- | --------------------------------------------                                                   |
| host_id         | integer | body | **Required**. The host id to run the script on.                                                |

> | script_id       | integer | body | The script to run. The current contents of this saved script will be executed on the host. Only one of script_id and script_contents must be provided.  |

Reviewer: ```suggestion
| script_id       | integer | body | The ID of the existing saved script to run. Only one of either `script_id` or `script_contents` can be included in the request; omit this parameter if using `script_contents`.  |
```

> | script_contents | string  | body | The contents of the script to run. Only one of script_id and script_contents must be provided. |

Reviewer: ```suggestion
| script_contents | string  | body | The contents of the script to run. Only one of either `script_id` or `script_contents` can be included in the request; omit this parameter if using `script_id`. |
```

> 

Reviewer: Would add a note below the table clarifying what happens if you happen to include both `script_id` and `script_contents` (whether that's an error, or one being ignored). e.g.:
```suggestion
Note that if both `script_id` and `script_contents` are included in the request, this endpoint will respond with an error.
```
or

```suggestion
Note that if both `script_id` and `script_contents` are included in the request, `script_id` will be used and `script_contents` will be ignored.
```
(or whatever actually happens)

| Name            | Type    | In   | Description                                      |
| ----            | ------- | ---- | --------------------------------------------     |
| host_id         | integer | body | **Required**. The host id to run the script on.  |
| script_id       | integer | body | The script to run. The current contents of this saved script will be executed on the host. Only one of script_id and script_contents must be provided.  |

> | script_contents | string  | body | The contents of the script to run. Only one of script_id and script_contents must be provided. |

Reviewer: Same feedback as "Run script asynchronously" above

| Name            | Type    | In   | Description                                      |
| ----            | ------- | ---- | --------------------------------------------     |
| host_id         | integer | body | **Required**. The host id to run the script on.  |

> | script_id       | integer | body | The script to run. The current contents of this saved script will be executed on the host. Only one of script_id and script_contents must be provided.  |
> | script_contents | string  | body | The contents of the script to run. Only one of script_id and script_contents must be provided. |

Reviewer: Just duplicated the changes to the run asynchronously parameters for extra clarity (since folks won't necessarily look at these in order):
```suggestion
| script_id       | integer | body | The ID of the existing saved script to run. Only one of either `script_id` or `script_contents` can be included in the request; omit this parameter if using `script_contents`.  |
| script_contents | string  | body | The contents of the script to run. Only one of either `script_id` or `script_contents` can be included in the request; omit this parameter if using `script_id`. |
```

> Note: `exit_code` can be `null` if Fleet hasn't heard back from the host yet.

### Upload a script

_Available in Fleet Premium_

> Uploads a script, making it available to run on hosts of the specified team (or no team).

Reviewer: ```suggestion
Uploads a script, making it available to run on hosts assigned to the specified team (or no team).
```

`POST /api/v1/fleet/scripts`

#### Parameters

| Name            | Type    | In   | Description                                      |
| ----            | ------- | ---- | --------------------------------------------     |
| script          | file    | form | **Required**. The file containing the script.    |

> | team_id         | integer | form | The team id for the script. If specified, the script is applied to only hosts that are assigned to the specified team. If not specified, the profile is applied only to hosts not assigned to any team. |

Reviewer: Took a stab at shortening this. Also, would use the phrase "no team" to be consistent with what we call things in the UI.
```suggestion
| team_id         | integer | form | The team ID. If specified, the script will only be available to hosts assigned to this team. If not specified, the script will only be available to hosts on **no team**.  |
```

#### Example

`POST /api/v1/fleet/scripts`

##### Request headers

```http
Content-Length: 306
Content-Type: multipart/form-data; boundary=------------------------f02md47480und42y
```

##### Request body

```http
--------------------------f02md47480und42y
Content-Disposition: form-data; name="team_id"

1
--------------------------f02md47480und42y
Content-Disposition: form-data; name="script"; filename="myscript.sh"
Content-Type: application/octet-stream

echo "hello"
--------------------------f02md47480und42y--

```

##### Default response

`Status: 200`

```json
{
  "script_id": 1227
}
```

### Delete a script

_Available in Fleet Premium_

Deletes an existing script.

`DELETE /api/v1/fleet/scripts/{id}`

#### Parameters

| Name            | Type    | In   | Description                                           |
| ----            | ------- | ---- | --------------------------------------------          |

> | id              | integer | path | **Required**. The identifier of the script to delete. |

Reviewer: ```suggestion
| id              | integer | path | **Required**. The ID of the script to delete. |
```
